### ISSUE TITLE
large message bubble backgrounds fail to render

### ISSUE THREAD

**Issue by 5boro** [WhisperSystems/TextSecure#3148]
Message doesn't show when sent and has no background when received.

![screenshot_2015-05-10-18-18-34](https://cloud.githubusercontent.com/assets/3460479/7555114/04a6a6ac-f742-11e4-8741-a836870ea78d.png)
![screenshot_2015-05-10-18-18-49](https://cloud.githubusercontent.com/assets/3460479/7555115/04c40dbe-f742-11e4-9f62-0ac58a022c02.png)

Its seems to be specific to this piece of text, although it could be length related:

```
Rêve :
- Big fish de Tim Burton
"un homme raconte si souvent ses histoires qu'il ne finit par ne faire plus qu'un avec ses histoires et ainsi il devient immortel"

- "Utopie" de Thomas More 

- Gustave Thibon "Faire rêver les hommes est le plus sûr moyen de les tenir endormis" 

- Jeanne Bernis "L'imagination" le manque, l'éloignement rend désirable, la cristallisation, le rêve est trompeur et engendre des désillusions, sublime la realite

- Freud : "Le bonheur est un rêve d'enfant réalisé à l'âge adulte
" Tout rêve est réalisation de désir"
"Le rêve ne pense ni ne calcule, d'une manière générale il ne juge pas, il se contente de transformer"

- Paulo Coelho :" C'est la possibilité de réaliser un rêve qui rend la vie interessante"

- Vera Wong : "Princess" - Jeune fille habillé en princesse pour un parfum, fait rêver 

- Raymond Trousson : "Le ver est dans le fruit" ,  monde uniforme, tout est contrôlé, révèle l'anti-utopie 

- Le seigneur des anneaux 


Objet : 
- Antoine de Saint-Exupery : "Ce n'est point dans l'objet que réside le sens des choses mais dans la démarche" 

- Fight Club de D.Fincher- 1999 - Scene de la rêverie autour du catalogue Ikea

- Gran Torino -  Clint Eastwood -  2008 - que léguer et à qui léguer. 

- Émile Zola "L'assomoir" - La machine et l'homme dans l'industrie, personnification de l'objet (Marteau) 

- Wall-E- Andrew Stanton 

- Alexandro Zinna "L'objet et ses interfaces"- une différence entre l'objet utile(uniforme, usine) et l'objet esthétique (unique, artisanal). Les deux tendent parfois à se rapprocher 

-" Le bonheur c'est de continuer à désirer ce que l'on possède" Saint Augustin 

- Marie Piquemal "Ils ont décidé de vivre avec cent objets"-Liberation
```

Device info: HTC one M7 GE, stock android 5.0.1, TextSecure 2.14.1
Debug log: https://gist.github.com/anonymous/5d33a6dac7f6873495fd

**Comment by 2-4601** [comment]
Tried to reproduce this with the given text using different themes and languages but failed.
TS 2.14.1
stock Android 5.1, Nexus 4
Android 4.0.4, Huawei U8800

I don't get any of those `qdmemalloc ` or `OpenGLRenderer` warnings.

**Comment by mcginty** [comment]
Seems like this can be caused when rendering large TextViews - http://stackoverflow.com/questions/27444387/issue-on-openglrenderer-path-too-large-to-be-rendered-to-a-texture. Might need to set the layerType to `software` to mitigate, because this is a new device so I doubt it's a legacy problem.

**Comment by 5boro** [comment]
I can reproduce it on a LG g2, Cyanogen/Stock Android 5.0.1, TextSecure 2.14.1

**Comment by mcginty** [comment]
@5boro is this a consistent thing, or just with long texts?

**Comment by 5boro** [comment]
It is consistent, but seems specific to this text.
Debug log for the LG g2 : https://gist.github.com/anonymous/d45a79c0b0a1c469436b

**Comment by 5boro** [comment]
I just noticed that the text in written in white over white background, so the bug is just that there's no background either when sent or received with this text.

![](http://zippy.gfycat.com/GreedyHandyHydatidtapeworm.gif)

**Comment by agrajaghh** [comment]
might be related to #3134?

**Comment by 5boro** [comment]
Just tried a few things, it is length related, and the limit seems to be between 750 and 1000 characters.
The text was over 1700.

**Comment by McLoo** [comment]
It's not the number of chars, but the number of lines!
If the background is visible for you, just feed some/many/big amount of line breaks and the background gets veeery transparent.

Tested on dark theme. no blue bubble background, but visible white characters on black background.
As on dark theme sender and receiver text color is white <> light theme sender:white, receiver:black

**Comment by 5boro** [comment]
Good catch, but wasn't @2-4601 able to reproduce it?

**Comment by 2-4601** [comment]
@5boro I'm **unable** to reproduce this :)
The text has 39 lines. Even if paste it twice it works without issues on both of my devices with both themes.

**Comment by McLoo** [comment]
@5boro cause copying your text sample here eliminates all the line breaks, when pasted into TS - At least when I was testing.

**Comment by 2-4601** [comment]
I saved the text (with all 39 lines) to a file and copypasted it to TextSecure from there.

**Comment by 5boro** [comment]
Are you on Android >5 @McLoo ?
Because I had the occasion to test it on a few of my friends phone, 2 Galaxy S5 and a Meizu MX3, all were on Kitkat, and the bug wasn't reproducing.

**Comment by 2-4601** [comment]
I can reproduce this now on my Nexus 4, stock Lollipop 5.1.
It requires 107 line breaks (I had also one character/byte on each line).

Here's a [debug log](https://gist.github.com/b8ae65d214bc99bc22f4) where I
* send a message with 106 line breaks and it renders correctly
* check that the log contains no `OpenGLRenderer` warnings
* send another message with **107** line breaks and the rendering breaks

**Comment by mcginty** [comment]
Thanks for the testing, guys. My guess is that the line-number threshold to set off this error is variable depending on what the device decides is its max OpenGL texture size.

**Comment by mcginty** [comment]
Yeah, any bubble > 4096px will cause this error, whether we set the LayerType to software or hardware.

Even more strangely, if we use a ColorDrawable or RectShape drawable instead of a GradientDrawable or RoundedRectShape drawable, this error doesn't happen. It appears Android doesn't handle rounded corners efficiently, making me wonder if we're better off using nine-patches instead of shape drawables for them.

**Comment by mcginty** [comment]
Basically we need to not use our own shape-based drawables and move to 9patches. I started the process, but there's a lot of busywork for asset generation and re-aligning stuff.

**Comment by mcginty** [comment]
Fixed in 2.21.0
